### PR TITLE
[KYUUBI #527] Support test with MiniYARNCluster

### PR DESCRIPTION
**Pull request by turboFei**
### _Why are the changes needed?_

Support test with mini yarn cluster.

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-main/pom.xml [hunk 1]
```diff
@@ -138,6 +138,13 @@
             <artifactId>${iceberg.name}</artifactId>
             <scope>test</scope>
         </dependency>
+
+        <dependency>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-client-minicluster</artifactId>
+            <scope>test</scope>
+        </dependency>
+
```
Comment (turboFei) [inline kyuubi-main/pom.xml:147]: will remove this blank line

File: kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithMiniYarnCluster.scala [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+      } else {
+        coreSiteProperties += k -> v
+      }
+    }
+
+    writeXmlFile("core-site.xml", coreSiteProperties.toMap)
+    writeXmlFile("hdfs-site.xml", hdfsSiteProperties.toMap)
+    writeXmlFile("yarn-site.xml", yarnSiteProperties.toMap)
+    writeXmlFile("mapred-site.xml", mapRedProperties.toMap)
+  }
```
Comment (turboFei) [inline kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithMiniYarnCluster.scala:157]: save these xml files in case that we need set HADOOP_CONF_DIR

**Comment by turboFei** [comment]
trying to add unit test for it.

**Comment by turboFei** [comment]
I think I need combine withYarnMiniCluster and withKyuubiServer

File: kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala [hunk 3]
```diff
@@ -19,22 +19,31 @@ package org.apache.kyuubi.operation
...
   private var server: KyuubiServer = _
   private val metastore = Utils.createTempDir()
 
+  override protected val startMiniYarnCluster: Boolean = false
+
   override def beforeAll(): Unit = {
+    super.beforeAll()
+    if (startMiniYarnCluster) {
+      yarnClusterConfig().foreach { case (k, v) =>
+        conf.set(s"spark.hadoop.${k}", v)
```
Comment (turboFei) [inline kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala:43]: Is it the correct way to integrate with mini yarn cluster?

If not, have we transfer env property like `HADOOP_CONF_DIR` into kyuubi server?

Comment (turboFei) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala:43]: cc @yaooqinn

File: kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithMiniYarnCluster.scala [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+      yarnConf.setInt("yarn.scheduler.capacity.root.default.capacity", 100)
+      yarnConf.setFloat("yarn.scheduler.capacity.root.default.user-limit-factor", 1)
+      yarnConf.setInt("yarn.scheduler.capacity.root.default.maximum-capacity", 100)
+      yarnConf.set("yarn.scheduler.capacity.root.default.state", "RUNNING")
+      yarnConf.set("yarn.scheduler.capacity.root.default.acl_submit_applications", "*")
+      yarnConf.set("yarn.scheduler.capacity.root.default.acl_administer_queue", "*")
+      yarnConf.setInt("yarn.scheduler.capacity.node-locality-delay", -1)
+
+      // Set RM bind host to localhost to avoid java.net.BindException
+      yarnConf.set("yarn.resourcemanager.bind-host", "localhost")
```
Comment (turboFei) [inline kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithMiniYarnCluster.scala:91]: try set rm bind host to localhost to avoid java.net.BindException

Comment (turboFei) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithMiniYarnCluster.scala:91]: set more `bind-host` to localhost.

yarn.nodemanager.bind-host
yarn.web-proxy.bind-host
yarn.router.bind-host

Refer: https://github.com/apache/hadoop/blob/26b8f678b2af98a5812c2d116e724159e37c72dd/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/resources/yarn-default.xml

File: kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala (hunk 3 above)
Comment (turboFei) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala:43]: It does not work, all application id are local spark application id.

Comment (turboFei) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala:43]: missed to set master to yarn, let me retry

**Comment by yaooqinn** [comment]
will take a look soon, thanks for driving this

**Comment by turboFei** [comment]
I will try to fix it.
convert it to draft temporarily.

File: kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala [hunk 5]
```diff
@@ -19,22 +19,33 @@ package org.apache.kyuubi.operation
...
+import org.apache.kyuubi.Utils
 import org.apache.kyuubi.config.KyuubiConf
 import org.apache.kyuubi.config.KyuubiConf._
+import org.apache.kyuubi.config.internal.Tests.TESTING_YARN_CONF_DIR
 import org.apache.kyuubi.ha.HighAvailabilityConf.{HA_ZK_ACL_ENABLED, HA_ZK_QUORUM}
 import org.apache.kyuubi.server.KyuubiServer
 import org.apache.kyuubi.zookeeper.{EmbeddedZookeeper, ZookeeperConf}
 
-trait WithKyuubiServer extends KyuubiFunSuite {
+trait WithKyuubiServer extends WithMiniYarnCluster {
```
Comment (ulysses-you) [inline kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala:30]: We don't need to make all test extend `WithMiniYarnCluster `. We can choose to use it if some suites need.

File: kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala [hunk 6]
```diff
@@ -19,22 +19,33 @@ package org.apache.kyuubi.operation
...
   private val metastore = Utils.createTempDir()
 
+  override protected val startMiniYarnCluster: Boolean = false
+
   override def beforeAll(): Unit = {
+    super.beforeAll()
+    if (startMiniYarnCluster) {
+      conf.set("spark.master", "yarn")
+      conf.set("spark.yarn.queue", "root.default")
+      conf.set("spark.executors.instances", "1")
```
Comment (ulysses-you) [inline kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala:45]: Does these configs really used ? Kyuubi config cann't pass to Spark as far as I know.

Comment (turboFei) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala:45]: spark.master need to be set, it could be transferred to spark sql engine.

File: kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala (hunk 5 above)
Comment (turboFei) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala:30]: will try it later

File: kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala (hunk 6 above)
Comment (ulysses-you) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala:45]: ah you are right, the related code is in `SparkProcessBuilder`. 

The config `conf.set("spark.master", "yarn")` seems necessary, but the others config should use `setIfMissing` instead of `set` ?

**Comment by turboFei** [comment]
GA passed.
travis.ci failed because of log length.

**Comment by ulysses-you** [comment]
`Kyuubi / Build (-Pspark-3.1 -Pkyuubi-extension-spark_3.1)` test failed.
```
*** 1 TEST FAILED ***
ANTLR Tool version 4.7 used for code generation does not match the current runtime version 4.8ANTLR Tool version 4.7 used for code generation does not match the current runtime version 4.8
```

Not sure why this failed, can you rebase to master ?

File: kyuubi-main/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala [hunk 7]
```diff
@@ -150,7 +150,14 @@ trait ProcBuilder {
...
     }
     error match {
-      case UNCAUGHT_ERROR => KyuubiSQLException(
+      case UNCAUGHT_ERROR =>
+        if (Utils.isTesting) {
+          val log = new String(Files.readAllBytes(engineLog.toPath), Charset.forName("UTF-8"))
+          // scalastyle:off println
+          println(s"Please check this engine log: $log")
+          // scalastyle:on println
+        }
```
Comment (ulysses-you) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala:159]: Do we need this ?

Comment (turboFei) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala]: Just used for debug, have removed it

File: pom.xml [hunk 8]
```diff
@@ -1082,6 +1082,13 @@
                 <version>${ldapsdk.version}</version>
             </dependency>
 
+            <dependency>
+                <groupId>org.apache.hadoop</groupId>
+                <artifactId>hadoop-client-minicluster</artifactId>
+                <version>${hadoop.version}</version>
+                <scope>test</scope>
```
Comment (yaooqinn) [inline pom.xml:1089]: no scope needed here

File: kyuubi-main/src/test/scala/org/apache/kyuubi/WithMiniYarnCluster.scala [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+      if (System.nanoTime() - startTimeNs > TimeUnit.SECONDS.toNanos(10)) {
+        throw new IllegalStateException("Timed out waiting for RM to come up.")
+      }
+      debug("RM address still not set in configuration, waiting...")
+      TimeUnit.MILLISECONDS.sleep(100)
+    }
+
+    info(s"RM address in configuration is ${config.get(YarnConfiguration.RM_ADDRESS)}")
+
+    assert(hadoopConfDir.isDirectory())
```
Comment (ulysses-you) [inline kyuubi-main/src/test/scala/org/apache/kyuubi/WithMiniYarnCluster.scala:110]: We can move assert to the top of this method.

File: kyuubi-main/src/test/scala/org/apache/kyuubi/operation/KyuubiOperationYarnClusterSuite.scala [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  override val hadoopConfDir = Utils.createTempDir().toFile
+
+  override protected val conf: KyuubiConf = {
+    KyuubiConf().set(ENGINE_INIT_TIMEOUT, 300000L)
+      .set("spark.master", "yarn")
+      .set(TESTING_HADOOP_CONF_DIR, hadoopConfDir.getAbsolutePath)
+      .setIfMissing("spark.executor.instances", "1")
+  }
+
+  test("KYUUBI #527- Support test with mini yarn cluster") {
```
Comment (ulysses-you) [inline kyuubi-main/src/test/scala/org/apache/kyuubi/operation/KyuubiOperationYarnClusterSuite.scala:38]: Can we add some test for yarn only? e.g. check some yarn config from sql. `SELECT "${spark.app.id}" as id;`

**Review by ulysses-you — COMMENTED** [review]
@turboFei thank you. This is really helpful.

File: kyuubi-main/src/test/scala/org/apache/kyuubi/WithMiniYarnCluster.scala [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    File.createTempFile("token", ".txt", hadoopConfDir)
+    saveHadoopConfFiles(yarnClusterConfig)
+  }
+
+  override def afterAll(): Unit = {
```
[the remark this answers is not in the document]
Comment (turboFei) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/WithMiniYarnCluster.scala]: Have refactored the code. thanks

File: kyuubi-main/src/test/scala/org/apache/kyuubi/server/MiniYarnService.scala [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    yarnConfig.set(s"hadoop.proxyuser.$currentUser.groups", "*")
+    yarnConfig.set(s"hadoop.proxyuser.$currentUser.hosts", "*")
+    yarnConfig
+  }
+
+  override def initialize(conf: KyuubiConf): Unit = {
+    hadoopConfDir = Utils.createTempDir().toFile
+    yarnConf = newYarnConfig()
+    yarnCluster = new MiniYARNCluster(name, 1, 1, 1)
+    yarnCluster.init(yarnConf)
```
Comment (ulysses-you) [inline kyuubi-main/src/test/scala/org/apache/kyuubi/server/MiniYarnService.scala:72]: We don't need to write yarn config into `hadoopConfDir ` ?

File: kyuubi-main/src/test/scala/org/apache/kyuubi/server/MiniYarnService.scala [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    yarnCluster.start()
+    val config = yarnCluster.getConfig()
+    val startTimeNs = System.nanoTime()
+    while (config.get(YarnConfiguration.RM_ADDRESS).split(":")(1) == "0") {
+      if (System.nanoTime() - startTimeNs > TimeUnit.SECONDS.toNanos(10)) {
+        throw new IllegalStateException("Timed out waiting for RM to come up.")
+      }
+      debug("RM address still not set in configuration, waiting...")
+      TimeUnit.MILLISECONDS.sleep(100)
+    }
```
Comment (yaooqinn) [inline kyuubi-main/src/test/scala/org/apache/kyuubi/server/MiniYarnService.scala:86]: eventually -> config.get(YarnConfiguration.RM_ADDRESS).split(":")(1) == "0"?

File: kyuubi-main/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala [hunk 14]
```diff
@@ -27,16 +27,26 @@ import org.apache.hadoop.security.UserGroupInformation
...
-    override val env: Map[String, String] = sys.env)
+    envMap: Map[String, String] = sys.env)
   extends ProcBuilder with Logging {
 
   import SparkProcessBuilder._
 
+  override protected def env: Map[String, String] = {
+    val testingHadoopConfDir = conf.getOption(Tests.TESTING_HADOOP_CONF_DIR.key)
+    if (Utils.isTesting && testingHadoopConfDir.isDefined) {
+      envMap ++ Map("HADOOP_CONF_DIR" -> testingHadoopConfDir.get)
```
Comment (yaooqinn) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:44]: can we set HADOOP_CONF_DIR in pom.xml to a dummy dir like `project.build`?

File: kyuubi-main/src/test/scala/org/apache/kyuubi/server/MiniYarnService.scala (hunk 12 above)
Comment (turboFei) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/server/MiniYarnService.scala:72]: yes, I have transferred all the hadoop config into kyuubi server

File: kyuubi-main/src/test/scala/org/apache/kyuubi/server/MiniYarnService.scala (hunk 13 above)
Comment (turboFei) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/server/MiniYarnService.scala:86]: Will refactor it.

File: kyuubi-main/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala (hunk 14 above)
Comment (turboFei) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:44]: I remembered I tried it before, let me try again.

Comment (turboFei) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:44]: It does not work, it can not be transferred to process launched by command.

Comment (yaooqinn) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:44]: how about we pass environments to `SparkProcessBuilder`, as it's designed to support this

Comment (yaooqinn) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:44]: Can we create a separate PR first to make it work?

Comment (turboFei) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:44]: sure, will create a separate pr to support transfer env map.

File: kyuubi-main/src/test/scala/org/apache/kyuubi/WithKyuubiServerWithMiniYarnService.scala [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+  override def beforeAll(): Unit = {
+    miniYarnService = new MiniYarnService()
+    miniYarnService.initialize(new KyuubiConf(false))
+    miniYarnService.start()
+
+    KyuubiHadoopUtils.toSparkPrefixedConf(miniYarnService.getHadoopConf()).foreach { case (k, v) =>
+      conf.set(k, v)
+    }
+    conf.set(TESTING_HADOOP_CONF_DIR, miniYarnService.getHadoopConfDir())
```
Comment (ulysses-you) [inline kyuubi-main/src/test/scala/org/apache/kyuubi/WithKyuubiServerWithMiniYarnService.scala:45]: If this `hadoopDirConf` is empty and we transform configs to kyuubi config, what's the meaning of `HADOOP_CONF_DIR`. Do I missed anything ?

Comment (yaooqinn) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/WithKyuubiServerWithMiniYarnService.scala:45]: just a hardcode check at spark side

Comment (ulysses-you) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/WithKyuubiServerWithMiniYarnService.scala:45]: ah I see.

**Review by ulysses-you — APPROVED** [review]

**Comment by turboFei** [comment]
thanks, merged to master
